### ISSUE TITLE
[REVIEW]: GridapDistributed: a massively parallel finite element toolbox in Julia

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4157]
**Submitting author:** @amartinhuertas (<a href="http://orcid.org/0000-0001-5751-4561">Dr. Alberto F. Martin</a>)
**Repository:** <a href="https://github.com/gridap/GridapDistributed.jl/" target ="_blank">https://github.com/gridap/GridapDistributed.jl/</a>
**Version:** v0.2.4
**Editor:** @Kevin-Mattheus-Moerman
**Reviewer:** @PetrKryslUCSD, @Leticia-maria, @jedbrown
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/3bd4bf42932beb69c49ebd4f88db45c2/status.svg)](https://joss.theoj.org/papers/3bd4bf42932beb69c49ebd4f88db45c2)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/3bd4bf42932beb69c49ebd4f88db45c2"><img src="https://joss.theoj.org/papers/3bd4bf42932beb69c49ebd4f88db45c2/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/3bd4bf42932beb69c49ebd4f88db45c2/status.svg)](https://joss.theoj.org/papers/3bd4bf42932beb69c49ebd4f88db45c2)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/gridap/GridapDistributed.jl/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@PetrKryslUCSD & @Leticia-maria & @jedbrown, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @Kevin-Mattheus-Moerman know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @PetrKryslUCSD
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/gridap/GridapDistributed.jl/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/gridap/GridapDistributed.jl/">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@amartinhuertas) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @Leticia-maria
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/gridap/GridapDistributed.jl/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/gridap/GridapDistributed.jl/">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@amartinhuertas) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @jedbrown
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/gridap/GridapDistributed.jl/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/gridap/GridapDistributed.jl/">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@amartinhuertas) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @PetrKryslUCSD, @Leticia-maria, @jedbrown it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.04 s (1178.8 files/s, 104503.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                           30            547            154           2627
TOML                             4            110              1            459
Markdown                        12             97              0            226
YAML                             2              0              6            110
Bourne Shell                     1              1              3              3
-------------------------------------------------------------------------------
SUM:                            49            755            164           3425
-------------------------------------------------------------------------------


Statistical information for the repository '35bcbf2144fe0509a1393d8a' was
gathered on 2022/02/13.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
PDF failed to compile for issue #4157 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by Kevin-Mattheus-Moerman** [comment]
@PetrKryslUCSD, @Leticia-maria, @jedbrown, @amartinhuertas this is where the review takes places. The reviewers may now start their review, guided by the check lists at the top of those issue.
Reviewers can provide feedback here or open dedicated issues on the projects repository and link to them here.

**Comment by jedbrown** [comment]
@whedon generate pdf from branch joss_paper

**Comment by jedbrown** [comment]
@whedon check references from branch joss_paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss_paper
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1137/141000671 is OK
- 10.21105/JOSS.02520 is OK
- 10.1137/20M1328786 is OK
- 10.1007/978-3-030-23957-2 is OK
- 10.5194/GMD-6-2099-2013 is OK
- 10.1137/100791634 is OK
- 10.1006/JCPH.1996.0047 is OK
- 10.3402/TELLUSA.V56I5.14436 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04157/joss.04157/10.21105.joss.04157.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04157/joss.04157/10.21105.joss.04157.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
👋 @Leticia-maria - @whedon is a bot, now in the processing of being replaced by @editorialbot.  Your editor is @Kevin-Mattheus-Moerman, and he will be able to help you submit review comments,

**Comment by whedon** [comment]
:wave: @Leticia-maria, please update us on how your review is going (this is an automated reminder).

**Comment by Kevin-Mattheus-Moerman** [comment]
@arfon it looks like whedon still does the reminders. :point_up:

**Comment by arfon** [comment]
> @arfon it looks like whedon still does the reminders. ☝️

Yeah, that will stop being the case over the next ~2 weeks. Apologies for any confusion caused.

**Comment by Kevin-Mattheus-Moerman** [comment]
@PetrKryslUCSD, @Leticia-maria, @jedbrown I am just checking in to see how things are going. Could you please provide an update on review progress? If you need more time that is okay too. Thanks again for your help!

**Comment by PetrKryslUCSD** [comment]
@whedon generate pdf from branch joss_paper

**Comment by PetrKryslUCSD** [comment]
@editorialbot generate pdf from branch joss_paper

**Comment by PetrKryslUCSD** [comment]
@editorialbot generate pdf

**Comment by Kevin-Mattheus-Moerman** [comment]
@PetrKryslUCSD thanks for your help here! Are you able to tick that last box? The license box is un-ticked. Note the MIT license used is OSI approved and accepted for JOSS.

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot set joss_paper as branch

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot generate pdf

**Comment by PetrKryslUCSD** [comment]
Sorry, I overlooked that.

**Comment by Leticia-maria** [comment]
I would like to be a permanent reviewer of JOSS, is it possible?

**Comment by jedbrown** [comment]
Sorry about my delays. I'll be able to do this in a week.

**Comment by Kevin-Mattheus-Moerman** [comment]
@jedbrown Sounds good. Thanks.

**Comment by Kevin-Mattheus-Moerman** [comment]
@jedbrown @Leticia-maria can you provide an update on review progress? Thanks!

**Comment by Kevin-Mattheus-Moerman** [comment]
@amartinhuertas thanks for your comment. Our apologies for the delays here. It has been very difficult across the board (since the COVID pandemic), for all JOSS submissions, to find reviewers and also to get them to commit to a timely review. We've been forced to be more flexible regarding time as a consequence. Having said that I think here it has been taking a bit long for them to respond so I have just contacted the reviewers separately just now again, and hopefully we'll hear from them shortly. 

In the unlikely case the reviewers become unresponsive then yes we can find different reviewers, however lets wait to give them a change to get back to us. Thanks for your patience.

**Comment by jedbrown** [comment]
I'm still here and will get to this soon. I've been buried in end-of-semester rush, a bunch of reporting deadlines, and sick kids. I'm very sorry about the delays.

**Comment by Leticia-maria** [comment]
Hello, @Kevin-Mattheus-Moerman: All issues addressed. I strongly recommend publishing as it is.

**Comment by Leticia-maria** [comment]
Sorry for the delays, I was in my last semester of my Graduation

**Comment by Leticia-maria** [comment]
I am really excited to review another papers

**Comment by amartinhuertas** [comment]
Hi @Kevin-Mattheus-Moerman, 

do u think it might be possible to go ahead with two reviews given the difficulties to get the third review?

Thanks!
@amartinhuertas

**Comment by jedbrown** [comment]
Sorry about my delays (a medical emergency last week laid waste to all my plans). Thanks for this cool package.

## Project docs
1. I know it's hard to address disparate audiences, but I think it's pretty unclear to someone used to invoking mpiexec themselves what the workflow looks like. Could you address such users somewhere discoverable, including instructions for running a *.jl demo (in parallel) rather than mainly pointing to notebook workflows?
1. I'm honestly surprised the shapes broadcast here. In any case, this isn't the Jacobian of the previous line, which represents a regularized p=4 Laplacian. https://github.com/gridap/GridapDistributed.jl/blob/master/test/PLaplacianTests.jl#L27

## Paper
1. "compact syntax, that": remove comma
1. use @key instead of [@key] for [in-text citation](https://pandoc.org/MANUAL.html#extension-citations), as with "see (Verdugo & Badia, 2021) for a detailed overview" (and subsequent occurrences)
1. the code sample will render correctly if you put it in a fenced Julia block, and that way will use a font that matches the other literals in the text. I know JOSS failed to handle unicode fonts properly in older iterations, but this was fixed a few months ago.
1. Not a problem per se, but I'm surprised to see the same time step used for the 256 (approx 40 km) and 512 (approx 20 km) elements. I'd expect to step at a constant CFL (otherwise you won't resolve phase). With a 480s time step, this would resolve waves on the order of 80 m/s versus 40 m/s on the two meshes, respectively. The latter is slower than the jet stream. Also, I'm not sure what to think of the ringing artifacts in the (low resolution) figure 3a.
1. Many references have have improper capitalization (APA says to use sentence case)
1. Gibson et al has missing journal info
1. Discussion of related packages is slim; I'd suggest at least discussing relative scope of FEniCS and perhaps something like FreeFEM++ and MFEM/deal.II/libMesh.

**Comment by amartinhuertas** [comment]
Hi @jedbrown, 

thanks for your review. 

I have pushed a new version of the article which I believe addresses all your concerns. Let me know if this is not the case.

Let me answer below inline to each comment. 

> I know it's hard to address disparate audiences, but I think it's pretty unclear to someone used to invoking mpiexec themselves what the workflow looks like. Could you address such users somewhere discoverable, including instructions for running a *.jl demo (in parallel) rather than mainly pointing to notebook workflows?

Yes. See https://github.com/gridap/GridapDistributed.jl#execution-modes-and-how-to-execute-the-program-in-each-mode

> I'm honestly surprised the shapes broadcast here. In any case, this isn't the Jacobian of the previous line, which represents a regularized p=4 Laplacian. https://github.com/gridap/GridapDistributed.jl/blob/master/test/PLaplacianTests.jl#L27

I am not sure which is the source of the confusion (if there is any), nor why this is not the Jacobian of the previous line. Can you please elaborate a little bit more? Please note that `∇du` represents the gradient applied to the trial basis, while  `∇u` the gradient applied to a single FE function which respresents the current nonlinear solver iterate. 

1. Addressed in the new version of the article.
2. Addressed in the new version of the article.
3. I tried the Julia fenced block but still Unicode characterers are not rendered appropiately. I am used the Docker image supplied with joss to render the paper in my local machine. 
4. Addressed in the new version of the article.
5. Addressed in the new version of the article.
6. It is a book. I have fixed the bib entry.
7. Addressed in the new version of the article.

@amartinhuertas

**Comment by amartinhuertas** [comment]
@editorialbot generate pdf

**Comment by amartinhuertas** [comment]
@editorialbot generate pdf

**Comment by jedbrown** [comment]
Regarding the 4-Laplacian, the current code is
```julia
  σ(∇u) =(1.0+∇u⋅∇u)*∇u
  dσ(∇du,∇u) = 2*∇u⋅∇du + (1.0+∇u⋅∇u)*∇du
```
The first term of the second line is a scalar while the second is a vector. In Julia, we normally can't add these without `.+` broadcast; e.g., `1 + [2,3]` is an error and you need `1 .+ [2,3]`.

I usually write the derivative like this
```julia
  dσ(∇du,∇u) = (2*∇u⋅∇du)*∇u + (1.0+∇u⋅∇u)*∇du
```
or something like this to really expose the geometric intuition that Newton linearization yields a symmetric rank-1 modification to the isotropic effective diffusivity.
```julia
  μ(∇u) = 1 + ∇u⋅∇u
  σ(∇u) = μ(∇u) * ∇u
  dσ(∇du,∇u) = (μ(∇u)*I + 2 * ∇u * ∇u') * ∇du
```
(In 3D, $p>2$ elongates it like a pencil while $p<2$ squashes it like a pancake because the second term has a negative multiplier. The problem is singular in the $p=1$ limit and degenerate in the $p=∞$ limit.)

> I tried the Julia fenced block but still Unicode characterers are not rendered appropiately. I am using the Docker image supplied with joss to render the paper in my local machine.

Hmm, I ran like this and it seems okay?
```julia
$ docker run -it --rm -v $(pwd):/data -u $(id -u):$(id -g) openjournals/inara -o pdf,crossref joss_paper/paper.md
```
<details>

![image](https://user-images.githubusercontent.com/3303/170489803-e65fe302-0f93-47d9-9400-570409638d66.png)
```diff
diff --git i/joss_paper/paper.md w/joss_paper/paper.md
index 8f8a49d..ff3998a 100644
--- i/joss_paper/paper.md
+++ w/joss_paper/paper.md
@@ -55,7 +55,31 @@ In order to confirm our previous claims on expressiveness, conciseness and produ
 (In order to fully understand the code snippet, familiarity with the high level API of Gridap is assumed.)
 The domain is discretized using the parallel Cartesian-like mesh generator built-in in GridapDistributed. The only minimal burden posed on the programmer versus Gridap is a call to the `prun` function of PartitionedArrays right at the beginning of the program. With this function, the programmer sets up the PartitionedArrays communication backend (i.e., MPI communication backend in the example), specifies the number of parts and their layout (i.e., `(2,2)` 2D layout in the example), and provides a function (using Julia do-block syntax for function arguments in the example) to be run on each part. This function is equivalent to a sequential Gridap script, except for the `CartesianDiscreteModel` call, which, in GridapDistributed, also requires the `parts` argument passed back by the `prun` function. In a typical cluster environment, this example would be executed on 4 MPI tasks from a terminal as `mpirun -n 4 julia --project=. example.jl`.
 
-![](code.pdf)
+```julia
+using Gridap
+using GridapDistributed
+using PartitionedArrays
+partition = (2,2)
+prun(mpi,partition) do parts
+  domain = (0,1,0,1)
+  mesh_partition = (4,4)
+  model = CartesianDiscreteModel(parts,domain,mesh_partition)
+  order = 2
+  u((x,y)) = (x+y)^order
+  f(x) = -Δ(u,x)
+  reffe = ReferenceFE(lagrangian,Float64,order)
+  V = TestFESpace(model,reffe,dirichlet_tags="boundary")
+  U = TrialFESpace(u,V)
+  Ω = Triangulation(model)
+  dΩ = Measure(Ω,2*order)
+  a(u,v) = ∫( ∇(v)·∇(u) )dΩ
+  l(v) = ∫( v*f )dΩ
+  op = AffineFEOperator(a,l,U,V)
+  uh = solve(op)
+  writevtk(Ω,"results",cellfields=["uh"=>uh,"grad_uh"=>∇(uh)])
+end
+```
+<!--  ![](code.pdf) -->
 
 
 # Parallel scaling benchmark
```

</details>

**Comment by amartinhuertas** [comment]
> I usually write the derivative like this

Ok. I know see what you mean. You are absolutely correct. I wrote the derivative a thousand times on paper and I was sure that it matched the code, but in the code there was a missing multiplication by `∇u` in the first term! Pair programming is definitely efficient, I think I would have not been able to see this error in 1000 years.

The code snippet below reproduces the situation. The last assert fails meaning that, as one would expect, `2*∇u⋅∇du` is NOT equivalent to `(2*∇u⋅∇du)*∇u`. In regards to why it broadcasts (i.e., why no error is produced), the cause is that operations among Gridap `CellField`s (e.g., `∇u` and `∇du` are `CellField`s) have broadcasting semantics by construction. Once an operation tree among `CellField`s is built,  this tree is then evaluated at a set of integration points during numerical integration. The operations among tensors of floating point numbers which are triggered are broadcast-like operations. 

```julia
using Gridap

domain = (0,1,0,1)
cells = (4,4)
model = CartesianDiscreteModel(domain,cells)
Ω  = Interior(model)
dΩ = Measure(Ω,2)

reffe = ReferenceFE(lagrangian,Float64,1)
V = TestFESpace(model,reffe,dirichlet_tags="boundary")
U = TrialFESpace(V)

du  = get_trial_fe_basis(U)
∇du = gradient(du)

u = FEFunction(U,rand(num_free_dofs(U)))
∇u = gradient(u)

op1=2*∇u⋅∇du
op2=(1.0+∇u⋅∇u)*∇du
op3=(2*∇u⋅∇du)*∇u
x=get_cell_points(dΩ.quad)

op1(x)[1]
op2(x)[1]
op3(x)[1]
@assert (op1+op2)(x)[1] == op1(x)[1] .+ op2(x)[1]
@assert (op2+op3)(x)[1] == (op1+op2)(x)[1]
```

**Comment by amartinhuertas** [comment]
> Hmm, I ran like this and it seems okay?

Ok. I tried your patch + suggested command and now it renders correctly. Thanks! I would swear that the command I was using was taken from the JOSS instructions per authors. But I now see that https://joss.readthedocs.io/en/latest/submitting.html?highlight=docker#docker does not match the command I was using. Nevermind, it works :-)

**Comment by jedbrown** [comment]
Yeah, JOSS switched to the new software stack (which contained https://github.com/openjournals/inara/pull/4 to handle unicode fonts, plus PDF/A support and other benefits) a couple days after you first submitted. So no mistake on your end, but we can use the new system now (and it'll have better accessibility).

**Comment by Kevin-Mattheus-Moerman** [comment]
@jedbrown how are you, can you provide an update on review progress? Thanks again for your help!

**Comment by jedbrown** [comment]
@editorialbot generate pdf

**Comment by jedbrown** [comment]
@amartinhuertas Are you going to update the p-Laplacian test?

I think the FEniCS book (https://fenicsproject.org/book/) is a better reference. I think this needs some rewording because it seems the point is that FEniCS *does* rely on a (reasonably sophisticated) form compiler that is problematic to apply to various challenging problems. For example, return mapping algorithms in plasticity need to be written in C++ with bits specific to dolfin/firedrake, thus losing many of the benefits of working with UFL in the first place. The pragmatic need for such escape hatches is why the likes of [PyOP2](https://github.com/OP2/PyOP2) were created.

> A hallmark of the latter ones is that they are able to achieve a very expressive and compact (yet efficient) syntax without relying on a sophisticated compiler of variational forms (see, e.g., @Kirby2006). To this end, this high level syntax is transformed into low-level code using the Julia JIT compiler. Apart from a more intrincate workflow (e.g., a Python front-end and a C/C++ back-end), one of the limitations of form compilers is that they are typically quite rigid systems not designed to be extended by average users.

**Comment by jedbrown** [comment]
@editorialbot check references

**Comment by jedbrown** [comment]
Also, the libMesh paper is https://doi.org/10.1007/s00366-006-0049-3

With these minor revisions, I'm happy to recommend for publication.

**Comment by amartinhuertas** [comment]
> @amartinhuertas Are you going to update the p-Laplacian test?

Yes, I already updated it 12 days ago. https://github.com/gridap/GridapDistributed.jl/pull/88/files

Am I missing something else here? (i.e., another source file with the same error?)

**Comment by amartinhuertas** [comment]
> Am I missing something else here? (i.e., another source file with the same error?)

Ok, I think I know now what happened. I updated the master branch, but you were looking at the code in the joss_paper branch. I have now pull master into joss_paper branch so that both branches are synced.

**Comment by amartinhuertas** [comment]
> I think this needs some rewording

I have reworded. Let me know if you are happy with the sentence in the current form.

**Comment by amartinhuertas** [comment]
> Also, the libMesh paper is https://doi.org/10.1007/s00366-006-0049-3

Done!

**Comment by amartinhuertas** [comment]
> I think the FEniCS book (https://fenicsproject.org/book/) is a better reference. 

Done!

**Comment by amartinhuertas** [comment]
@jedbrown ... I am done with my revision changes.

**Comment by jedbrown** [comment]
@editorialbot generate pdf

**Comment by jedbrown** [comment]
> There are a number of high quality open source parallel finite element packages available in the literature. Some examples are deal.II [@dealII93], libMesh [@libMeshPaper], MFEM [@mfem], FEMPAR [@Badia2017], FEniCS [@fenics-book], or FreeFEM++ [@freefem], to name a few. All these  packages have their own set of features, potentials, and limitations. Among these, FEniCS and FreeFEM++ are perhaps the closest ones in scope and spirit to the packages in the Gridap ecosystem. A hallmark of the latter ones compared to the former ones is that their very expressive and compact (yet efficient) syntax is transformed into low-level code using the Julia JIT compiler and thus they do not need a sophisticated compiler of variational forms nor a more intrincate workflow (e.g., a Python front-end and a C/C++ back-end).

1. This seems to be saying that a hallmark of FEniCS/FreeFEM++ relative to deal.II et al is that it uses Julia JIT. I think your intent is perhaps that a hallmark of Gridap relative to these is that you implement weak forms in a full-featured language (Julia) and thus benefit from Julia's JIT rather than a custom one.
2. "intrincate" -> "intricate"

**Comment by amartinhuertas** [comment]
@editorialbot generate pdf

**Comment by amartinhuertas** [comment]
@editorialbot check references

**Comment by Kevin-Mattheus-Moerman** [comment]
> Looks good. @Kevin-Mattheus-Moerman wave feel free to move forward.

Great thanks @jedbrown!

**Comment by amartinhuertas** [comment]
> The review started off for version v0.2.4, can you confirm if this version is still accurate or if we need to update the version tag for this submission?

The latest public release of GridapDistributed.jl is v0.2.6. We need to update the version tag for this submission to v0.2.6.

> Can you please archive a copy of the reviewed work on ZENODO? You can do this manually if you like. Some users find [these automated steps](https://docs.github.com/en/repositories/archiving-a-github-repository/referencing-and-citing-content) useful.

Done. 

> Please ensure that the ZENODO title and authors match those of the paper.

Done.

> Can you please list the ZENODO archive DOI here?

10.5281/zenodo.6622081

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot set version to v0.2.6

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot set v0.2.6 as version

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot set 10.5281/zenodo.6622081 as archive

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot recommend accept

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot check references

**Comment by Kevin-Mattheus-Moerman** [comment]
@openjournals/dev hmmm can you check what is going on here? :point_up:

**Comment by xuanxu** [comment]
The [code for a couple of figures](https://github.com/gridap/GridapDistributed.jl/blob/joss_paper/joss_paper/paper.md?plain=1#L91-L95) with multiple images with the same caption was invalid. [This PR](https://github.com/gridap/GridapDistributed.jl/pull/90) changes it to use a single image with multiple panels and using [valid syntax](https://joss.readthedocs.io/en/latest/submitting.html#example-paper-and-bibliography).

**Comment by Kevin-Mattheus-Moerman** [comment]
@amartinhuertas can you check this :point_up: 

Thanks for your help @xuanxu!

**Comment by Kevin-Mattheus-Moerman** [comment]
@xuanxu this does not explain the reference checking to behave differently though right? :point_up:

**Comment by amartinhuertas** [comment]
> @amartinhuertas can you check this point_up

I already accepted the PR. Thanks @xuanxu !

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot recommend-accept
